python/tskit/metadata.py

```diff
@@ -120,6 +123,191 @@ def decode(self, data: bytes) -> bytes:
+            if do_sort:
+                return collections.OrderedDict(items)
+            else:
+                return dict(items)
+        elif isinstance(obj, list) or isinstance(obj, tuple):
+            return [cls.order_by_index(v, False) for v in obj]
+        else:
+            return obj
+
+    def __init__(self, schema: Mapping[str, Any]) -> None:
+        # This codec requires that all properties are required and additional ones
+        # not allowed. Rather than get schema authors to repeat that everywhere
+        # we add it here
+        def enforce_fixed_properties(obj):
+            if type(obj) == list:
+                return [enforce_fixed_properties(j) for j in obj]
+            elif type(obj) == dict:
+                ret = {k: enforce_fixed_properties(v) for k, v in obj.items()}
+                if ret.get("type") == "object":
+                    ret["required"] = list(ret.get("properties", {}).keys())
+                    ret["additionalProperties"] = False
+                return ret
+            else:
+                return obj
+
+        schema = enforce_fixed_properties(schema)
+        try:
+            StructCodecSchemaValidator.check_schema(schema)
+        except jsonschema.exceptions.SchemaError as ve:
+            raise exceptions.MetadataSchemaValidationError(str(ve)) from ve
+        self.schema = self.order_by_index(schema)
+
+        def make_encode(sub_schema):
+            """
+            Create a function that can encode objects of this schema
+            """
+            if sub_schema["type"] == "array":
+                element_encoder = make_encode(sub_schema["items"])
+                return lambda array: struct.pack("<Q", len(array)) + b"".join(
+                    element_encoder(ele) for ele in array
+                )
+            elif sub_schema["type"] == "object":
+                sub_encoders = {
+                    key: make_encode(prop)
+                    for key, prop in sub_schema["properties"].items()
+                }
+                return lambda obj: b"".join(
+                    sub_encoder(obj[key]) for key, sub_encoder in sub_encoders.items()
+                )
+            elif sub_schema["type"] == "string":
+                return lambda string: struct.pack(
+                    "<" + sub_schema["binaryFormat"], string.encode()
+                )
+            elif sub_schema["type"] == "null":
+                return lambda _: b""
+            else:
+                return struct.Struct("<" + sub_schema["binaryFormat"]).pack
+
+        def make_decode(sub_schema):
+            """
+            Create a function that can decode objects of this schema
+            """
+            if sub_schema["type"] == "array":
+                element_decoder = make_decode(sub_schema["items"])
+                array_length_size = struct.calcsize("<Q")
+
+                def array_decode(bytes_, offset):
```

Comment:
Just a comment really. I find ``bytes_`` a bit awkward and would tend to use ``buff`` or ``buffer`` for variables like this. No biggie either way.

Reply:
Buffer is better, renamed.